pkg/prometheus/operator.go

```diff
@@ -0,0 +1,591 @@
+// Copyright 2023 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/blang/semver/v4"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/relabel"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/assets"
+	"github.com/prometheus-operator/prometheus-operator/pkg/informers"
+	"github.com/prometheus-operator/prometheus-operator/pkg/k8sutil"
+	"github.com/prometheus-operator/prometheus-operator/pkg/operator"
+)
+
+// OperatorInterface is the interface shared by Server and Agent operators.
+type OperatorInterface interface {
+	GetLogger() log.Logger
+	GetServiceMonitorInformers() *informers.ForResource
+	GetPodMonitorInformers() *informers.ForResource
+	GetProbeInformers() *informers.ForResource
+	GetNamespaceInformers() cache.SharedIndexInformer
+	GetMetrics() *operator.Metrics
+}
+
+func SelectServiceMonitors(ctx context.Context, po OperatorInterface, p monitoringv1.PrometheusInterface, store *assets.Store) (map[string]*monitoringv1.ServiceMonitor, error) {
```

Comment:
I would create a struct and turn the Select* functions into receiver's methods.

````golang
type ResourceSelector struct {
    l log.Logger
    store *assets.Store
    podMonitorLister ListAllByNamespacer
    serviceMonitorLister ListAllByNamespacer
    ...
}

type ListAllByNamespacer interface {
    ListAllByNamespace(namespace string, selector labels.Selector, appendFn cache.AppendFunc) error
}

func NewResourceSelector(l log.Logger, p monitoringv1.PrometheusInterface, store *assets.Store, podMonitorLister ListAllByNamespacer, ...) *ResourceSelector {
    ...
}

func (rs *ResourceSelector) SelectServiceMonitors(ctx context.Context) (map[string]*monitoringv1.ServiceMonitor, error) {
    ...
}
````

Reply:
Nice idea!

Could you clarify what is the purpose of the interface `ListAllByNamespace`, not sure if I'm lacking context, but it is not obviously clear what it is used for 🤔